### PR TITLE
Update tracer to guard for a nil ctx

### LINKED ISSUE #3109: Creating a new Span with a nil context causes a panic
### Description

Creating a new Span with a nil context causes a panic
Example https://go.dev/play/p/SXEK1EX_GzQ

Stack trace:

```
goroutine 1 [running]:
context.WithValue({0x0?, 0x0?}, {0x58aac0?, 0x6151e0?}, {0x5b9a20?, 0xc0002d4600?})
	/usr/local/go-faketime/src/context/context.go:525 +0x178
go.opentelemetry.io/otel/trace.ContextWithSpan(...)
	/tmp/gopath1305294289/pkg/mod/go.opentelemetry.io/otel/trace@v1.9.0/context.go:25
go.opentelemetry.io/otel/sdk/trace.(*tracer).Start(0xc00001e8c0?, {0x0?, 0x0?}, {0x5bf3e3?, 0x5?}, {0x0?, 0x60?, 0x726800?})
	/tmp/gopath1305294289/pkg/mod/go.opentelemetry.io/otel/sdk@v1.9.0/trace/tracer.go:58 +0x28c
main.main()
	/tmp/sandbox3842089424/prog.go:35 +0x264
```

IMO Otel should guard for this and if it's nil fill in a `context.Background()`. PR to do that: https://github.com/open-telemetry/opentelemetry-go/pull/3110

### Environment

- OS: Linux
- Architecture: x64
- Go Version: 1.19
- opentelemetry-go version: 1.9.0 (based on panic stack trace)

### Steps To Reproduce

1. Provide a nil context to `Start()` (see Go Playground link in description)

### Expected behavior
Typical Otel behavior seems to either No-op or use reasonable defaults. While this panic isn't coming from the Otel lib, but rather an std lib, I'd still expect the sdk to - at worst - provide a NoOp span based on similar behavior in other parts

### PR DESCRIPTION
**Pull request by Cull-Methi**
Updates the tracer Start to guard on a nil context. Closes #3109

This might be better inside the `trace.ContextWithSpan` call on L63 - happy to change this to there if preferred

### CODE REVIEW

**Comment by hanyuancheung** [comment]
> [CLA Not Signed](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/9522512/187086161/3110/#/?version=2)
> 
> * [❌](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/9522512/187086161/3110/#/?version=2) - login: @Cull-Methi / name: Mitch Usher . The commit ([1d2a936](https://github.com/open-telemetry/opentelemetry-go/commit/1d2a936cfdf429519b396f3d9c8715d54d79dfaf), [97d898c](https://github.com/open-telemetry/opentelemetry-go/commit/97d898cc8513989baf26bea6b1781405c2140111)) is not authorized under a signed CLA. [Please click here to be authorized](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/9522512/187086161/3110/#/?version=2). For further assistance with EasyCLA, [please submit a support request ticket](https://jira.linuxfoundation.org/servicedesk/customer/portal/4).

@Cull-Methi Pls sign the CLA.

**Comment by Cull-Methi** [comment]
> > [CLA Not Signed](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/9522512/187086161/3110/#/?version=2)
> > 
> > * [❌](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/9522512/187086161/3110/#/?version=2) - login: @Cull-Methi / name: Mitch Usher . The commit ([1d2a936](https://github.com/open-telemetry/opentelemetry-go/commit/1d2a936cfdf429519b396f3d9c8715d54d79dfaf), [97d898c](https://github.com/open-telemetry/opentelemetry-go/commit/97d898cc8513989baf26bea6b1781405c2140111)) is not authorized under a signed CLA. [Please click here to be authorized](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/9522512/187086161/3110/#/?version=2). For further assistance with EasyCLA, [please submit a support request ticket](https://jira.linuxfoundation.org/servicedesk/customer/portal/4).
> 
> @Cull-Methi Pls sign the CLA.

Just did :)

File: sdk/trace/trace_test.go [hunk 1]
```diff
@@ -363,6 +363,21 @@ func TestStartSpanWithParent(t *testing.T) {
...
+	alwaysSampleTp := NewTracerProvider()
+	sampledTr := alwaysSampleTp.Tracer("AlwaysSampled")
+
+	defer func() {
+        if r := recover(); r != nil {
+            t.Error("unexpected panic creating span with nil context")
+        }
+    }()
+
+	_, _ := sampledTr.Start(nil, "should-not-panic")
```
Comment (hanyuancheung) [inline sdk/trace/trace_test.go:378]:
```suggestion
	_, _ = sampledTr.Start(nil, "should-not-panic")
```
no new variable on the left of := , replace with =.

**Review by hanyuancheung — APPROVED** [review]

File: sdk/trace/trace_test.go [hunk 2]
```diff
@@ -363,6 +363,21 @@ func TestStartSpanWithParent(t *testing.T) {
...
+	alwaysSampleTp := NewTracerProvider()
+	sampledTr := alwaysSampleTp.Tracer("AlwaysSampled")
+
+	defer func() {
+        if r := recover(); r != nil {
+            t.Error("unexpected panic creating span with nil context")
+        }
+    }()
+
+	_, _ = sampledTr.Start(nil, "should-not-panic")
```
Comment (hanyuancheung) [inline sdk/trace/trace_test.go:378]:
```suggestion
	_, _ = sampledTr.Start(context.TODO(), "should-not-panic")
```

Comment (Cull-Methi) [reply sdk/trace/trace_test.go]: I think this doesn't help us here as `TODO()` returns a non-nil context, so the test isn't doing what we expect - https://go.dev/play/p/i9aqgGPU7PV (compared to https://go.dev/play/p/SXEK1EX_GzQ)

I think we'll need to instead just disable lint on that particular line

**Review by dashpole — APPROVED** [review]

File: sdk/trace/trace_test.go [hunk 3]
```diff
@@ -363,6 +363,21 @@ func TestStartSpanWithParent(t *testing.T) {
...
+func TestStartSpanWithNilContext(t *testing.T) {
+	alwaysSampleTp := NewTracerProvider()
+	sampledTr := alwaysSampleTp.Tracer("AlwaysSampled")
+
+	defer func() {
+		if r := recover(); r != nil {
+			t.Error("unexpected panic creating span with nil context")
+		}
+	}()
+	//nolint // no nil context, but that's the point of the test
```
Comment (MadVikingGod) [inline sdk/trace/trace_test.go:377]:
```suggestion
	//nolint:staticcheck // no nil context, but that's the point of the test
```

Lets only disable the one lint rule.

**Review by MadVikingGod — APPROVED** [review]
Add a changelog entry please.

**Review by XSAM — CHANGES_REQUESTED** [review]
I am not sure about this. Go Context does not have the convention to take nil context. For instance, [`Context.WithValue`](https://pkg.go.dev/context#WithValue) panics when the parameter is nil. https://cs.opensource.google/go/go/+/refs/tags/go1.19:src/context/context.go;l=523

If we accept nil context, that probably means we need to accept nil context everywhere to keep consistency since our end-users might think we can tolerate nil context, thus passing nil context everywhere.

**Comment by dashpole** [comment]
Because of [this](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/error-handling.md#basic-error-handling-principles), I don't think we can allow nil contexts to panic: `OpenTelemetry implementations MUST NOT throw unhandled exceptions at run time.`

I think that means we need to check for nil context everywhere.

**Review by XSAM — APPROVED** [review]

File: sdk/trace/trace_test.go (hunk 3 above)
Comment (Cull-Methi) [reply sdk/trace/trace_test.go]: Thanks! I wasn't sure which rule it fell under 😅

File: sdk/trace/trace_test.go [hunk 4]
```diff
@@ -363,6 +363,21 @@ func TestStartSpanWithParent(t *testing.T) {
...
+	tp := NewTracerProvider()
+	tr := tp.Tracer("NoPanic")
+
+	defer func() {
+		if r := recover(); r != nil {
+			t.Error("unexpected panic creating span with nil context")
+		}
+	}()
+	//nolint:staticcheck // no nil context, but that's the point of the test
+	tr.Start(nil, "should-not-panic")
```
Comment (MrAlias) [inline sdk/trace/trace_test.go:378]:
```suggestion
	// nolint:staticcheck // no nil context, but that's the point of the test.
	assert.NotPanics(t, func() { tr.Start(nil, "should-not-panic") })
```

File: sdk/trace/tracer.go [hunk 5]
```diff
@@ -37,6 +37,11 @@ var _ trace.Tracer = &tracer{}
 func (tr *tracer) Start(ctx context.Context, name string, options ...trace.SpanStartOption) (context.Context, trace.Span) {
 	config := trace.NewSpanStartConfig(options...)
 
+	// If ctx is nil, set to context.Background() as context.WithValue will panic on a nil value.
+	if ctx == nil {
+		ctx = context.Background()
```
Comment (MrAlias) [inline sdk/trace/tracer.go:42]:
```suggestion
	if ctx == nil {
		// Prevent trace.ContextWithSpan from panicking.
		ctx = context.Background()
```

**Review by MrAlias — COMMENTED** [review]
This looks good outside of minor issues.

**Review by MrAlias — APPROVED** [review]

**Pull request merged**